[mono] Fix static virtual calls to generic methods from gshared code

BrzVlad opened the pull request:
When the resolved method from the vtable is generic, we need to inflate it.

For example, in the gshared method we are calling on `Interface.InterfaceMethod<int> ()` (the type argument can either be fixed or determined based on the gshared method's context), and this method is resolved to generic method `Class.InterfaceMethod<T> ()`. We then need to inflate the method and resolve `T` to `int`.

BrzVlad committed:
[mono] Fix static virtual calls to generic methods from gshared code

When the resolved method from the vtable is generic, we need to inflate it.

For example, in the gshared method we are calling on `Interface.InterfaceMethod<int> ()` (the type argument can either be fixed or determined based on the gshared method's context), and this method is resolved to generic method `Class.InterfaceMethod<T> ()`. We then need to inflate the method and resolve `T` to `int`.

lambdageek reviewed (approved):
Is there an existing test we can enable?

BrzVlad commented:
@lambdageek This PR together with https://github.com/dotnet/runtime/pull/66350 should fix all runtime tests within `Loader/classloader/StaticVirtualMethods/GenericContext/`

BrzVlad committed:
Enable tests

BrzVlad committed:
[mono][interp] Fix calls to static virtual methods via delegates

BrzVlad committed:
[mono][aot] Fix constrained + ldftn

BrzVlad committed:
[mono][gsharing] Fix constrained + ldftn from gshared method

We can'd determine at compile time the static method if the constrained_class is a generic parameteri. We defer the computation of the method at runtime, in a similar fashion with calls.

BrzVlad commented:
/azp run runtime-extra-platforms

BrzVlad committed:
Disable test suite on android

It hits an issue in the android test runner

BrzVlad commented:
/azp run runtime-extra-platforms

BrzVlad commented:
/azp run runtime-extra-platforms

BrzVlad commented:
`runtime-extra-platforms` failures are unrelated
